### PR TITLE
Override the fvpatch field type basedon the FoamBC class

### LINKED ISSUE #126: Allow Hippo to override boundary condition types specified in `0/*`
## Description

Replace the `fvPatchFields` with ones representative of the FoamBCs internally 

## Context

At the moment, the type in 0/* has to match that corresponding to the FoamBC, it would be better to override it.

## Outputs

- All tests working with a system that overrides the values in `0/*`

### PR DESCRIPTION
**Pull request by mattfalcone1997**
## Summary

Changes `FoamBCBase` to set the correct fvPatchField when it is initialised

## Related Issue

Resolves #126 

## Checklist

- [x] Functionality implememented
  - [x] Make sure energy BCs updated if `T` is updated 
- [x] Some of the existing boundary condition tests updated to test the switch
  - [x] `laplace_fixed_gradient`, `fixed_value` and `mass_flow_rate`
- [x] Fixed bugs in timestepper class associated with the update  
- [x] Existing docs didn't mention need to set the correct underlying BC (as far as I can see)

### CODE REVIEW

File: include/bcs/FoamBCBase.h [hunk 1]
```diff
@@ -44,6 +47,16 @@ class FoamBCBase : public MooseObject, public Coupleable
   // Get the data vector of the MOOSE field on a subdomain
   std::vector<Real> getMooseVariableArray(int subdomain_id);
 
+  // Construct boundary patch for scalar fields
+  void constructFoamScalarPatch(const std::string & patch_name, const std::string & bc_type);
+
+  // update the energy equation, if the temperature field is updated the energy equation must be
```
Comment (k-collie) [inline include/bcs/FoamBCBase.h]:
```suggestion
  // Update the energy equation, if the temperature field is updated the energy equation must be
```

File: test/tests/bcs/fixed_value/foam/0/T [hunk 2]
```diff
@@ -27,8 +27,8 @@ boundaryField
     }
     right
     {
-        type            fixedValue;
-        value           uniform 0.;
+        type            fixedGradient;
+        gradient           uniform 0.;
```
Comment (k-collie) [inline test/tests/bcs/fixed_value/foam/0/T:32]: What would you think to reverting these and having a separate explicit test for this functionality?

Otherwise I think a comment would be useful here.

File: src/bcs/FoamBCBase.C [hunk 3]
```diff
@@ -25,6 +25,8 @@ FoamBCBase::validParams()
   params.addRequiredParam<std::string>("foam_variable",
                                        "Name of a Foam field. e.g. T (temperature) U (velocity).");
 
+  params.addPrivateParam<std::string>("_foam_bc_type");
```
Comment (k-collie) [inline src/bcs/FoamBCBase.C]: Could we get away without adding this private parameter?

If not could you add a comment explaining why its a parameter?

Would it also be possible to make it a `MooseEnum`?

**Review by k-collie — COMMENTED** [review]
First look, I need to look some more at the new methods in `FoamBCBase`

File: src/bcs/FoamBCBase.C (hunk 3 above)
Comment (mattfalcone1997) [reply src/bcs/FoamBCBase.C]: The alternative to making it a parameter is to make it a class member. I think `MooseEnum` could be a good choice, though. Which would you prefer?

File: test/tests/bcs/fixed_value/foam/0/T (hunk 2 above)
Comment (mattfalcone1997) [reply test/tests/bcs/fixed_value/foam/0/T:32]: I think  its okay to test these together but I will add a comment in the `tests` and `T` files

File: src/bcs/FoamBCBase.C (hunk 3 above)
Comment (k-collie) [reply src/bcs/FoamBCBase.C]: I think it should be a pure virtual method returning an `enum class` if possible, otherwise a class member. I think if we can avoid exposing it as a MOOSE parameter (unless necessary) that would be better encapsulation.

File: include/bcs/FoamBCBase.h [hunk 4]
```diff
@@ -7,16 +7,22 @@
...
+// valid underlying Foam BC types
+static MooseEnum _foam_bc_types("fixedGradient fixedValue");
 
 class FoamBCBase : public MooseObject, public Coupleable
 {
 public:
   static InputParameters validParams();
 
-  explicit FoamBCBase(const InputParameters & params);
+  explicit FoamBCBase(const InputParameters & params, const std::string & bc_type);
```
Comment (k-collie) [inline include/bcs/FoamBCBase.h]: Could you make this an `enum class` rather than a string? Then you could drop the runtime validity check

**Review by k-collie — COMMENTED** [review]
One more thing, still need to look at the big new methods too

File: include/bcs/FoamBCBase.h (hunk 4 above)
Comment (mattfalcone1997) [reply include/bcs/FoamBCBase.h]: I don't think we can do it easily because MooseEnum doesn't inherently know what the enum's contents are. If it were a MOOSE input parameter, the parameter system would check it's valid.

**Review by k-collie — APPROVED** [review]
Good stuff
